src/v/archival/ntp_archiver_service.cc

```diff
@@ -1823,6 +1833,248 @@ ss::future<> ntp_archiver::housekeeping() {
     }
 }
 
+ss::future<> ntp_archiver::apply_archive_retention() {
+    if (!may_begin_uploads()) {
+        co_return;
+    }
+
+    const auto& ntp_conf = _parent.get_ntp_config();
+    std::optional<size_t> retention_bytes;
+    std::optional<std::chrono::milliseconds> retention_ms;
+
+    if (ntp_conf.retention_bytes()) {
+        retention_bytes = ntp_conf.retention_bytes();
+    }
+
+    if (ntp_conf.retention_duration()) {
+        retention_ms = ntp_conf.retention_duration();
+    }
+
+    auto res = co_await _manifest_view->compute_retention(
+      retention_bytes, retention_ms);
```

Comment:
Would be good to skip replicating the command if the `archive_start_offset` result is equal to `archive_start_offset{model::offset{}, model::offset_delta{}}`. Currently, we'd just print a warn log when applying the command, but that can change.